fix: relax identifier limits

Shubhranshu153 opened the pull request:
## Description
Some of users of finch are reporting max length of 76 characters, which seems to stem from this validation
https://github.com/containerd/containerd/blob/92900bf7300b4872db2a43e82128fda9cde08a79/pkg/identifiers/validate.go#L58

Docker skips this validation via containerd library and does its own regex validation

## What changed
This change moves the validation function to nerdctl which would give greater control over it. This change modifies identifier validation for the following:
- volume create
- container create
- network create
- namestore
- composer new project
- buildargs secret

This brings these identifier compatible to docker but doesn't add checks for others. Docker seems to have additional checks, of which not all are required but some might be needed to be ported.

## Test
1. local unit test

AkihiroSuda commented on pkg/namestore/namestore.go:
```diff
@@ -48,8 +48,8 @@ type nameStore struct {
 func (x *nameStore) Acquire(name, id string) error {
-	if err := identifiers.Validate(name); err != nil {
-		return fmt.Errorf("invalid name %q: %w", name, err)
+	if err := validator.Validate(name); err != nil {
```
```suggestion
	if err := identifiers.ValidateDocker(name); err != nil {
```

AkihiroSuda replied:
The difference from containerd id validator has to be explained in godoc

Shubhranshu153 replied:
should we have a docker compat option in nerdctl config file in that way customer can configure that and the commands need not change.
Here i can have a check if the flag is set.

Shubhranshu153 replied:
i can add the flag as part of this pr.

Shubhranshu153 replied:
> The difference from containerd id validator has to be explained in godoc

will add an explanation in godoc.

AkihiroSuda replied:
> should we have a docker compat option in nerdctl config file in that way customer can configure that and the commands need not change.

Sounds overengineering

Shubhranshu153 replied:
ok then can just change the names and add a doc.

apostasie commented:
Hey @Shubhranshu153 

So, identifiers no longer have a length limit, is that correct?

Isn't this going to be problematic for anything that will end-up being used on the filesystem? Particularly Namestore etc.

I am not a windows expert, though some liminary googling suggests that anything over 256 chars might spell trouble.

Shubhranshu153 commented:
> Hey @Shubhranshu153
> 
> So, identifiers no longer have a length limit, is that correct?
> 
> Isn't this going to be problematic for anything that will end-up being used on the filesystem? Particularly Namestore etc.
> 
> I am not a windows expert, though some liminary googling suggests that anything over 256 chars might spell trouble.

yes that is correct, 
and yes windows has a 256 char limit and linux has 4095 for file paths, may be we can have a selective check based on OS. 
File path validation can be limited to mount points though. But for lets say container name same identifier may not be needed

pendo324 commented:
Just noting that the Windows file path length limit can be circumvented in recent versions ([source](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry)), but its opt-in for users

apostasie commented:
> > Hey @Shubhranshu153
> > So, identifiers no longer have a length limit, is that correct?
> > Isn't this going to be problematic for anything that will end-up being used on the filesystem? Particularly Namestore etc.
> > I am not a windows expert, though some liminary googling suggests that anything over 256 chars might spell trouble.
> 
> yes that is correct, and yes windows has a 256 char limit and linux has 4095 for file paths, may be we can have a selective check based on OS. File path validation can be limited to mount points though. But for lets say container name same identifier may not be needed

We might want to make the effort and move to using shas on the filesystem so that we can allow random identifiers, but that is likely a lot of work to get there and support existing stuff.

Meanwhile, I think we need to mindful about what identifiers we are relaxing, depending on whether they are going to end-up on the filesystem...

I think you mentioned some of (finch?) users were facing issues - so, is there a particular type of identifier (network name? image name? container name?) that is most problematic for them?

Shubhranshu153 commented:
@apostasie
have seen for volume names majorly. They tend to be longer, we can go the route of having less stringent rules for volume names and other stuffs we can keep the same and probably use 255 for file names. 

We can put it within a dockercompat flag to be more sure about it.

Shubhranshu153 commented:
May be the simplest thing would be to increase the limits of all to 255 as that seems to be able to support most of the use cases

apostasie commented:
> @apostasie have seen for volume names majorly. They tend to be longer, we can go the route of having less stringent rules for volume names and other stuffs we can keep the same and probably use 255 for file names.
> 
> We can put it within a dockercompat flag to be more sure about it.

@Shubhranshu153 

I just sent #3281 which suggests to change the VolumeStore to no longer use names as path components - which would remove the need for any limitations on volume names.
This is somewhat tricky to get right (specifically backward compatibility), but IMHO it is the only way to be able to really relax restrictions.
LMK what you think over there.

Shubhranshu153 commented:
@apostasie  replied on the thread. Additionally:
Also feel if the system is going to fail it, then let the system handle it. Earlier i think i had proposed the sha in the issue, but that complicates the way we want to recover the original string from it.

apostasie commented:
> @apostasie replied on the thread. 

Thanks.

> Additionally: Also feel if the system is going to fail it, then let the system handle it.

Well, we made the deliberate choice to store user-picked identifiers as path components - this is very much an internal implementation choice - and not dealing with failures would just leak that choice into the UX... I don't think it is a good thing.

> Earlier i think i had proposed the sha in the issue, but that complicates the way we want to recover the original string from it.

Not really for the VolumeStore.

I have not looked at what it would mean for Networks, or the NameStore (although I think the NameStore is problematic in its own right and we might need to reconsider it).

Shubhranshu153 commented:
not really for volume store -> from user perspective if i would want to list out the volumes available will i be able to get it from the sha?
 
From the above converstations it feels like we do want to keep a limit so probably 254 is a safe upper limit considering the system limits.  its not exactly docker compatible though

apostasie commented:
> not really for volume store -> from user perspective if i would want to list out the volumes available will i be able to get it from the sha?

Nothing changes from a user perspective (with the VolStore PR).
When you `list` volumes you see their names. You can inspect a volume by name. And you can see the name and the path of the volume.

As I suggested above, how we construct the path on the filesystem is an implementation detail IMHO that should be kept entirely orthogonal to UX.
This is true as well for the API we expose - nothing changes for the consumer.

> From the above converstations it feels like we do want to keep a limit so probably 254 is a safe upper limit considering the system limits. its not exactly docker compatible though

The containerd method `Validate` is specifically advertised as being safe to use for filesystem path components (although the length limitation is somewhat arbitrary and has more to do with editors max line length).

IMHO the only way to lift these restrictions is to change our implementation(s) where it matters so we no longer use identifiers as path components (like I did for the VolumeStore).

Short of doing that, and if we are to increase that length regardless, we need tests to match, and build confidence that it will work (clearly not only Windows has limitations on these).

Shubhranshu153 commented:
yeah length will be variable which is not ideal but we can have longer names and keeping the same checks. 
I feel any route we take which involves morphing identifiers would run into risks of legacy artifacts in general. Probably not worth the effort. Rather may be a good way will be to build standard around identifier length for each OS type and recommendations. If they are larger than 76 then we don't really need to think of backward compatibility.

apostasie commented:
> yeah length will be variable which is not ideal but we can have longer names and keeping the same checks. I feel any route we take which involves morphing identifiers would run into risks of legacy artifacts in general. Probably not worth the effort. Rather may be a good way will be to build standard around identifier length for each OS type and recommendations. If they are larger than 76 then we don't really need to think of backward compatibility.

Agreed that handling legacy user data is _generally_ a daunting proposition.

That being said, in the context of containers, I feel like things tend to be a lot more ephemeral and easy to just rebuild (especially named volumes) - furthermore, having to deal with legacy is a fact of life no matter what, and right now (with v2 around the corner) is probably the ideal time to do things like that.

On the general question of whether such a migration makes sense, I stand by my gun: using user-controlled-identifiers as-is as path components is a problematic design from the get go (regardless of the implementation imposing restrictions).

Anyhow, just shooting the shit :-).

Hope this is all helpful.

Shubhranshu153 commented:
Thats a fair point but handling the legacy would mean we need test for all the stuffs we want to support backward compatibility for. Or we can add an label to the identifier hash to suggest if it is something newly created, if that label doesnt exist we fallback to new workflow. Need to think for the different use cases what changes are needed

Having said this i still feel the more simpler workflow is via having configurable identifier lengths, i am not saying its a great way, may be this is something we can merge to the current stable version. The reason is this is a request from a few of current users of finch and also needed to support dev containers.

Shubhranshu153 commented:
considering that we closed the other pr, i would make the suggested changes probably with a limit of 255 .

Shubhranshu153 commented:
Added docs and changed the function name.
Didn't add the length constraint though as it will make it not really compatible with docker. 
An unintended change was min length, seems docker has a min length of 2 but for containerd it was 1, so that is not changed to 2. Added the description in the doc

AkihiroSuda commented on docs/validate.md:
```diff
@@ -0,0 +1,9 @@
+This document explains the changes made in the identifier validation logic between the containerd identifier implementation and the Docker compatible implementation `ValidateDockerCompat`.
```
This document should be moved to godoc of ValidateDockerCompat

Shubhranshu153 replied:
i am assuming the suggestion is to rename the file to: ValidateDockerCompat

Shubhranshu153 replied:
ah ok will do

AkihiroSuda commented on pkg/validator/validate.go:
```diff
@@ -0,0 +1,39 @@
+   limitations under the License.
+*/
+
+package validator
```
"validator" isn't descriptive; it doesn't tell what is validated

Shubhranshu153 replied:
added a change for identifiers

AkihiroSuda commented on pkg/validator/validate.go:
```diff
@@ -0,0 +1,39 @@
+   limitations under the License.
+*/
+
+package validator
```
```suggestion
package identifiers
```

Shubhranshu153 replied:
added a change for identifiers

Shubhranshu153 commented:
There was typo in the previous comment:

```
An unintended change was min length, seems docker has a min length of 2 but for containerd it was 1, so that is now changed to 2 to match docker. Added the description in the doc
```

AkihiroSuda commented on pkg/identifiers/validate.go:
```diff
@@ -0,0 +1,43 @@
+
+var AllowedIdentifierPattern = regexp.MustCompile(`^` + AllowedIdentfierChars + `+$`)
+
+// This function implements docker compatible identifier validation.
```
godoc comment begins with the function name

AkihiroSuda commented on pkg/mountutil/mountutil.go:
```diff
@@ -26,10 +26,10 @@ import (
 	"github.com/moby/sys/userns"
 	"github.com/opencontainers/runtime-spec/specs-go"
 
-	"github.com/containerd/containerd/v2/pkg/identifiers"
```
?

AkihiroSuda commented on pkg/namestore/namestore.go:
```diff
@@ -96,8 +95,8 @@ func (x *nameStore) Rename(oldName, id, newName string) error {
-	if err := identifiers.Validate(newName); err != nil {
-		return fmt.Errorf("invalid name %q: %w", oldName, err)
+	if err := identifiers.ValidateDockerCompat(newName); err != nil {
+		return fmt.Errorf("invalid new name: %w", err)
```
```suggestion
		return fmt.Errorf("invalid name %q: %w", newName, err)
```

Shubhranshu153 replied:
itention was not to repeat the identifier twice as the err also prints it, doesnt harm to be verbose, updated it.

Shubhranshu153 commented on cmd/nerdctl/network_create.go:
```diff
@@ -19,13 +19,12 @@ package main
+	"github.com/containerd/nerdctl/v2/pkg/identifiers"
 	"github.com/containerd/nerdctl/v2/pkg/strutil"
+
+	"github.com/spf13/cobra"
```
not sure why the linter changed the order, let me change back

Shubhranshu153 commented:
@AkihiroSuda 
Updated the code. Thanks for the suggestions.

apostasie commented on pkg/composer/serviceparser/build.go:
```diff
@@ -25,9 +25,9 @@ import (
-	"github.com/containerd/containerd/v2/pkg/identifiers"
 	"github.com/containerd/errdefs"
 	"github.com/containerd/log"
+	"github.com/containerd/nerdctl/v2/pkg/identifiers"
```
@Shubhranshu153 can you rebase? We just introduced some linting on imports.
Once rebased, you should be able to autotidy with `make lint-fix-imports`

AkihiroSuda replied:
Also `s/remove/relax/` in the commit message

Shubhranshu153 replied:
sure let me try that out

Shubhranshu153 replied:
updated it, the test seems to fail windows in setup can we retrigger.

apostasie replied:
Just ignore canary windows failures for now (#3306). There is a bug in the go setup action that prevents it from functioning with full go version ending in .0.

apostasie replied:
@Shubhranshu153 you may rebase to get CI failure fixed.

Shubhranshu153 replied:
rebased

apostasie commented on pkg/identifiers/validate.go:
```diff
@@ -0,0 +1,43 @@
+	"github.com/containerd/errdefs"
+)
+
+const AllowedIdentfierChars = `[a-zA-Z0-9][a-zA-Z0-9_.-]`
```
So, there is no length limit at all?

Shubhranshu153 replied:
No limit here, this seem to be what we have in docker, if we want a limit we can put 254 probably but i am not sure as that breaks compatibility

apostasie replied:
Fair. "Compat with docker" beats personal feelings :-)

AkihiroSuda commented on pkg/identifiers/validate.go:
```diff
@@ -0,0 +1,43 @@
+   limitations under the License.
+*/
+
+// This package implements functions for docker compatible identifier validation
```
godoc comment should be like "Package identifiers implements ..."
https://go.dev/blog/godoc

Shubhranshu153 replied:
updated the comments and the format

AkihiroSuda reviewed (approved):
Thanks
